File: services/user/src/services/redis.service.ts
/**
 * RedisService — cache get/set/invalidate operations for the user service.
 * Uses ioredis. Keys follow the pattern "user:{userId}" with 5-minute TTL.
 * See TDD Section 5.1 for caching strategy.
 */

import Redis from 'ioredis';
import { logger } from '../utils/logger';

export const redis = new Redis({
  host: process.env.REDIS_HOST ?? 'localhost',
  port: Number(process.env.REDIS_PORT ?? 6379),
  lazyConnect: true,
});

const DEFAULT_TTL_SECONDS = 300; // 5 minutes per TDD Section 5.1

export class RedisService {
  static async get<T>(key: string): Promise<T | null> {
    try {
      const raw = await redis.get(key);
      if (raw === null) return null;
      return JSON.parse(raw) as T;
    } catch (err) {
      logger.warn('Redis get failed', { key, error: String(err) });
      return null;
    }
  }

  static async set(key: string, value: unknown, ttlSeconds: number = DEFAULT_TTL_SECONDS): Promise<void> {
    try {
      await redis.set(key, JSON.stringify(value), 'EX', ttlSeconds);
    } catch (err) {
      logger.warn('Redis set failed', { key, error: String(err) });
    }
  }

  static async invalidate(...keys: string[]): Promise<void> {
    if (keys.length === 0) return;
    try {
      await redis.del(...keys);
    } catch (err) {
      logger.warn('Redis invalidate failed', { keys, error: String(err) });
    }
  }
}

export default RedisService;

File: services/pet/src/main/java/com/newcat/pet/service/RedisService.java
package com.newcat.pet.service;

/**
 * Redis cache service for pet profiles. Uses Spring Data Redis (Lettuce client).
 * Cache keys follow the pattern "pet:{petId}" with 5-minute TTL.
 * See TDD Section 5.1 for caching strategy.
 */

import com.fasterxml.jackson.databind.ObjectMapper;
import com.newcat.pet.util.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.Optional;

@Service
public class RedisService {

    @Autowired
    private StringRedisTemplate redisTemplate;

    private final ObjectMapper mapper = new ObjectMapper();
    private static final Logger logger = new Logger(RedisService.class);
    private static final Duration DEFAULT_TTL = Duration.ofMinutes(5);

    public <T> Optional<T> get(String key, Class<T> type) {
        try {
            String value = redisTemplate.opsForValue().get(key);
            if (value == null) {
                return Optional.empty();
            }
            return Optional.of(mapper.readValue(value, type));
        } catch (Exception e) {
            logger.error("Redis get failed", e, "key", key);
            return Optional.empty();
        }
    }

    public boolean setWithTTL(String key, Object value, int ttlSeconds) {
        try {
            String json = mapper.writeValueAsString(value);
            redisTemplate.opsForValue().set(key, json, Duration.ofSeconds(ttlSeconds));
            return true;
        } catch (Exception e) {
            logger.error("Redis set failed", e, "key", key);
            return false;
        }
    }

    public void delete(String key) {
        try {
            redisTemplate.delete(key);
        } catch (Exception e) {
            logger.error("Redis delete failed", e, "key", key);
        }
    }
}
